File: Scripts/CSharpScripts/Orbwlakers/OrbWalker.cs
using System.Numerics;
using Api;
using Api.Game.Calculations;
using Api.Game.GameInputs;
using Api.Game.Managers;
using Api.Game.Objects;
using Api.Game.ObjectTypes;
using Api.Menus;
using Api.Scripts;
using Api.Utils;
using Scripts.Utils;
using Timer = Scripts.Utils.Timer;

namespace Scripts.CSharpScripts.Orbwlakers;

public class OrbWalker : IOrbWalkScript
{
    public string Name => "OrbWalker";
    public ScriptType ScriptType => ScriptType.OrbWalker;
    public bool Enabled { get; set; }

    private readonly IScriptingState _scriptingState;
    private readonly IGameManager _gameManager;
    private readonly ILocalPlayer _localPlayer;
    private readonly IRenderer _renderer;
    private readonly IMinionSelector _minionSelector;
    private readonly IGameInput _gameInput;
    private readonly ITargetSelector _targetSelector;
    private readonly ITurretManager _turretManager;
    private readonly IRandomGenerator _randomGenerator;
    private readonly ISpellCaster _spellCaster;

    private readonly Timer _humanizerTimer;
    private readonly Timer _attackTimer;
    private readonly Timer _moveTimer;
    
    private readonly IToggle _blockAutoAttacks;
    private readonly IToggle _supportMode;
    private readonly IToggle _drawAttackRange;
    private readonly IToggle _drawKillableMinions;
    private readonly IToggle _humanizerSliderAddRandomDelay;
    private readonly IValueSlider _humanizerSlider;
    private readonly IValueSlider _pingSlider;
    private readonly IValueSlider _extraWindupSlider;
    private readonly IValueSlider _stoppingDistanceSlider;
    
    public OrbWalker(
        IMainMenu mainMenu,
        IScriptingState scriptingState,
        IGameManager gameManager,
        ILocalPlayer localPlayer,
        IRenderer renderer,
        IMinionSelector minionSelector,
        IGameInput gameInput,
        Timer humanizerTimer,
        Timer moveTimer,
        Timer attackTimer,
        ITargetSelector targetSelector,
        ITurretManager turretManager,
        IRandomGenerator randomGenerator,
        ISpellCaster spellCaster)
    {
        _scriptingState = scriptingState;
        _gameManager = gameManager;
        _localPlayer = localPlayer;
        _renderer = renderer;
        _minionSelector = minionSelector;
        _gameInput = gameInput;
        
        _humanizerTimer = humanizerTimer;
        _moveTimer = moveTimer;
        _attackTimer = attackTimer;
        _targetSelector = targetSelector;
        _turretManager = turretManager;
        _randomGenerator = randomGenerator;
        _spellCaster = spellCaster;

        var menu = mainMenu.CreateMenu(Name, ScriptType.OrbWalker);
        _humanizerSliderAddRandomDelay = menu.AddToggle("Humanizer random delay", true);
        _humanizerSlider = menu.AddFloatSlider("Humanizer", 75, 25, 300, 1, 0);
        _pingSlider = menu.AddFloatSlider("Ping", 35, 5, 300, 1, 0);
        _extraWindupSlider = menu.AddFloatSlider("Extra windup", 25, 0, 100, 1, 0);
        _stoppingDistanceSlider = menu.AddFloatSlider("Stopping distance", 70, 0, 250, 1, 0);

        _blockAutoAttacks  = menu.AddToggle("Block auto attacks", false);
        _supportMode = menu.AddToggle("Support mode", false);
        _drawAttackRange = menu.AddToggle("Draw attack range", true);
        _drawKillableMinions = menu.AddToggle("Draw killable minions", true);
    }
    
    public void OnLoad()
    {
    }

    public void OnUnload()
    {
    }
    
    private float GetAttackTime()
    {
        return MathF.Max(1.0f / _localPlayer.AttackSpeed, _pingSlider.Value / 1000);
    }

    private float GetWindupTime()
    {
        return (1.0f / _localPlayer.AttackSpeed) * _localPlayer.BasicAttackWindup + _extraWindupSlider.Value / 1000.0f;
    }
    
    private void MoveTo(Vector2 position)
    {
        if(!_moveTimer.IsReady) return;
        
        if (!_gameManager.GameCamera.WorldToScreen(_localPlayer.Position, out var playerScreenPosition))
        {
            return;
        }

        if (Vector2.Distance(playerScreenPosition, _gameInput.MousePosition) <= _stoppingDistanceSlider.Value)
        {
            return;
        }
        
        if (_gameInput.IssueOrder(position, IssueOrderType.Move))
        {
            var value = _humanizerSlider.Value;
            if (_humanizerSliderAddRandomDelay.Toggled)
            {
                value+=_randomGenerator.NextFloat(5, 50.0f);
            }
            _moveTimer.SetDelay(value / 1000.0f);
        }
    }

    private bool Attack(IAttackableUnit attackableUnit)
    {
        if ((_blockAutoAttacks.Toggled && attackableUnit.GameObjectType == GameObjectType.Hero) ||
            !_attackTimer.IsReady)
        {
            return false;
        }

        if (!_gameInput.Attack(attackableUnit)) return false;
        
        _attackTimer.SetDelay(GetAttackTime());
        _moveTimer.SetDelay(GetWindupTime());

        _humanizerTimer.SetDelay(_humanizerSlider.Value/1000);
        return true;

    }
    
    public void OnUpdate(float deltaTime)
    {
        if (_scriptingState.ActionType == ActionType.None || !_localPlayer.IsAlive)
        {
            return;
        }

        if (_spellCaster.IsCasting)
        {
            return;
        }
        
        if (_scriptingState.IsCombo)
        {
            Combo(deltaTime);
        }
        else if (_scriptingState.IsHaras)
        {
            Haras(deltaTime);
        }
        else if (_scriptingState.IsFarm)
        {
            Farm(deltaTime);
        }
        else if (_scriptingState.IsClear)
        {
            Clear(deltaTime);
        }
    }

    private void Combo(float deltaTime)
    {
        var target = _targetSelector.GetTarget();
        if (target is not null)
        {
            if (Attack(target))
            {
                return;
            }
        }
        MoveTo(_gameInput.MousePosition);
    }

    private void Haras(float deltaTime)
    {
        IAttackableUnit? target = GetKillableMinion();
        
        if (target is null)
        {
            target = _targetSelector.GetTarget();
        }
        
        if (target is not null)
        {
            if (Attack(target))
            {
                return;
            }
        }
        
        MoveTo(_gameInput.MousePosition);
    }
    
    private void Farm(float deltaTime)
    {
        var target = GetKillableMinion();
        if (target is not null)
        {
            if (Attack(target))
            {
                return;
            }
        }
        
        MoveTo(_gameInput.MousePosition);
    }

    private void Clear(float deltaTime)
    {
        IAttackableUnit? target = GetKillableMinion();

        if (target is null)
        {
            target = _turretManager.GetEnemyTurrets(_localPlayer.AttackRange).FirstOrDefault();
        }
        
        if (target is null)
        {
            target = _minionSelector.GetHealthiestMinion(_localPlayer.AttackRange);
        }
        
        if (target is null)
        {
            var monsters = _gameManager.ObjectManager.MonsterManager.GetMonsters(_localPlayer.AttackRange);
            target = monsters.MinBy(x => x.Health);
        }
        
        if (target is not null)
        {
            if (Attack(target))
            {
                return;
            }
        }
        
        MoveTo(_gameInput.MousePosition);
    }
    
    public void OnRender(float deltaTime)
    {
        if (_drawAttackRange.Toggled)
        {
            _renderer.CircleBorder3D(_localPlayer.Position, _localPlayer.AttackRange, Color.White, 2);
        }

        if (_stoppingDistanceSlider.Value > 50)
        {
            _renderer.CircleBorder3D(_localPlayer.Position, _stoppingDistanceSlider.Value, Color.Green, 2);
        }
        
        DrawKillableMinions();
    }
    
    private void DrawKillableMinions()
    {
        if (!_drawKillableMinions.Toggled)
        {
            return;
        }
        
        var range = MathF.Max(_localPlayer.AttackRange + 300, 1000);

        foreach (var minion in _minionSelector.GetKillableMinions(range).Select(x => x.Minion))
        {
            if(minion is null) continue;
            _renderer.CircleBorder3D(minion.Position, minion.CollisionRadius, Color.Red, 1);
        }
    }
    
    private IMinion? GetKillableMinion()
    {
        if (_supportMode.Toggled && _gameManager.HeroManager.GetAllyHeroes(1000).Any())
        {
            return null;
        }
        var target = _minionSelector.GetBestKillableMinion(_localPlayer.AttackRange);
        return target.Minion;
    }
}

File: T_T_PandorasBox/Resources/LuaScripts/Examples/Orbwalker.lua
local hero = Hero
local camera = GameCamera
local damageCalculator = DamageCalculator
local damagePrediction = DamagePrediction
local inputManager = InputManager
local gameInput = GameInput
local objectManager = ObjectManager
local minionManager = MinionManager
local monsterManager = MonsterManager
local plantManager = PlantManager
local wardManager = WardManager
local trapManager = TrapManager
local turretManager = TurretManager
local inhibitorManager = InhibitorManager
local missileManager = MissileManager

local time = 0
local humanizer = nil
local extraWindup = nil
local ping = nil
local stoppingDistance = nil

local humanizeTimer = 0
local attackTimer = 0
local moveTimer = 0

local comboHotkey = nil
local farmHotkey = nil
local clearHotkey = nil

local function updateTimers(deltaTime)
    humanizeTimer = humanizeTimer - deltaTime
    attackTimer = attackTimer - deltaTime
    moveTimer = moveTimer - deltaTime
end

local function getAttackTime()
    return math.max(1.0 / hero.AttackSpeed, ping.Value / 100)
end

local function getWindupTime()
    return (1.0 / getAttackTime()) * (hero.BasicAttackWindup / hero.AttackSpeedRatio)
end

local function move(position)
    if moveTimer > 0 then
        return
    end
    
    local screenPosition = camera:WorldToScreen(hero.Position)
    local mousePosition = gameInput.MousePosition

    if Math.Distance(screenPosition, mousePosition) <= stoppingDistance.Value then
        return
    end

    if gameInput:IssueOrder(position, IssueOrderType.Move) == true then
        humanizeTimer = humanizer.Value / 1000
    end
end

local function move2(position)
    if moveTimer > 0 then
        return
    end

    if gameInput:IssueOrder(position, IssueOrderType.Move) == true then
        humanizeTimer = humanizer.Value / 1000
    end
end

local function attackTarget(target)
    if attackTimer > 0 then
        return false
    end
    
    if gameInput:IssueOrder(target.Position, IssueOrderType.Attack) then
        attackTimer = getAttackTime()
        moveTimer = getWindupTime() + extraWindup.Value / 1000
        humanizeTimer = humanizer.Value / 1000

        return true
    end
    
    return false
end

local function getBestMinion()
    local minions = minionManager:GetEnemyMinions(hero.AttackRange)
    if minions == nil then
        return
    end

    local targetHp = 9999
    local target = nil
    for minion in luanet.each(minions) do
        local health = damagePrediction:PredictHealth(minion, 100)
        if minion ~= nil and health > 0 and health < targetHp then
            targetHp = minion.Health
            target = minion
        end
    end

    return target
end

local function getLasthitableMinion()
    local minion = getBestMinion()

    if minion ~= nil and minion.Health > damageCalculator:GetDamage(DamageType.Physical, hero, minion, hero.TotalAttackDamage) then
        return nil
    end
    return minion
end


function OnLoad()
    local menu = CreateMenu("Test Lua menu", ScriptType.OrbWalker)
    humanizer = menu:AddValueSlider("Humanizer", "Delay between actions is ms", 25, 25, 300)

    extraWindup = menu:AddValueSlider("Extra windup", "", 10, 0, 80)
    ping = menu:AddValueSlider("Ping", "", 35, 0, 250)

    stoppingDistance = menu:AddValueSlider("Stopping distance", "No move action will be taken when mouse is in range", 80, 0, 250)

    local hotkeysSubmenu = menu:AddSubMenu("Hotkeys", "You can setup hotkeys for orbwalker")
    
    comboHotkey = hotkeysSubmenu:AddHotkey("Combo", "Combo mode. Will attack enemy heroes.", VirtualKey.Spacebar, HotkeyType.Press)

    farmHotkey = hotkeysSubmenu:AddHotkey("Farm", "Farm mode. Will farm minions.", VirtualKey.X, HotkeyType.Press)

    clearHotkey = hotkeysSubmenu:AddHotkey("Clear", "Clear mode. Will attack minions.", VirtualKey.V, HotkeyType.Press)
    
end

function OnUnload()
end


function OnUpdate(deltaTime)

    updateTimers(deltaTime)

    if humanizeTimer > 0 then
        return
    end

    local target = nil
    local moveAction = false
    if clearHotkey.Enabled then
        target = getBestMinion()
        moveAction = true
    end

    if farmHotkey.Enabled then
        target = getLasthitableMinion()
        moveAction = true
    end

    if comboHotkey.Enabled then
        target = nil
        moveAction = true
    end

    if target ~= nil then
        if(attackTarget(target)) then
            return
        end
    end
    
    if moveAction == true then
        move(gameInput.MousePosition)
    end
end

function drawMinions()
    local minions = minionManager:GetEnemyMinions(hero.AttackRange)
    if minions == nil then
        return
    end

    for minion in luanet.each(minions) do
        if minion ~= nil then
            Renderer:Circle3D(minion.Position, minion.CollisionRadius, Color(0.0, 0.0, 1.0, 1.0), 1, time, 0.5, 1)
            local screenPosition = camera:WorldToScreen(minion.Position)
            --local damage = damageCalculator:GetDamage(DamageType.Physical, hero, minion, hero.TotalAttackDamage)
            --Renderer:Text(tostring(math.ceil(damage)) .. ' ' .. tostring(math.ceil(minion.Health)), screenPosition, 21, Color(1.0, 1.0, 1.0, 1.0))
        end
    end
end 

function OnRender(deltaTime)
    time = time + deltaTime
    return
    Renderer:Circle3D(hero.Position, hero.AttackRange, Color(0.0, 0.0, 1.0, 1.0), 1, time, 0.5, 2);
end
